### PR TITLE
Master ppapuga hibernate refactoring

### PR DESCRIPTION
**Pull request by PavleP**

### CODE REVIEW

File: src/main/java/eu/execom/labs/teamodoro/configuration/SpringWebInit.java [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
+import org.springframework.web.servlet.support.AbstractAnnotationConfigDispatcherServletInitializer;
+
+/**
+ * ServletInitializer class. Replacement for web.xml.
+ */
+public class SpringWebInit extends AbstractAnnotationConfigDispatcherServletInitializer {
+
+    @Override
+    protected Class<?>[] getRootConfigClasses() {
+        return new Class[] {SpringWebConfig.class};
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/configuration/SpringWebInit.java]: Should SpringWebConfig go into the root configuration or servlet configuration?

File: src/main/java/eu/execom/labs/teamodoro/entity/Team.java [hunk 2]
```diff
@@ -1,50 +1,48 @@
...
+        this.name = name;
+    }
+
+    public Collection<UserTeam> getUserTeams() {
+        return userTeams;
+    }
+
+    public void setUserTeams(List<UserTeam> userTeams) {
+        this.userTeams = userTeams;
+    }
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: Setter and getter are not symmetric in sense that getter returns a Collection, while setter is using a List. Please make this all use the same interface in the method signature, preferably a List, unless it is a recommended practice to use the Collection.

File: src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp [hunk 3]
```diff
@@ -55,71 +62,76 @@
...
+          <th>Users</th>
         </tr>
       </thead>
       <tbody>
         <c:forEach items="${teams}" var="team">
           <tr>
             <td><h4>${team.name}</h4>
               <button type="button" class="btn btn-danger btn-xs"
-                data-toggle="modal" data-target="#deleteTeamModal"><spring:message code="user.delete.team"/></button>
+                data-toggle="modal" data-target="#deleteTeamModal">Delete</button>
```
Comment (PavleP) [inline src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp]: Copy/paste of entire block from older project replaced already made i18n. This will be corrected at next commit.

File: src/main/java/eu/execom/labs/teamodoro/entity/Team.java [hunk 4]
```diff
@@ -1,50 +1,48 @@
...
+public class Team {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String name;
+
+    @OneToMany(mappedBy = "team")
+    private Collection<UserTeam> userTeams = new ArrayList<>();
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: Would Set be more efficient than a list/collection definition here?

Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: What's the preferred practice when mapping a collection?

File: src/main/java/eu/execom/labs/teamodoro/entity/User.java [hunk 5]
```diff
@@ -28,21 +27,23 @@
 
     private String email;
 
-    @ManyToMany
-    @JoinTable
-    private List<Team> teams;
+    @OneToMany(mappedBy = "user")
+    private Collection<UserTeam> userTeams = new ArrayList<>();
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/User.java]: Would Set be more efficient than a list/collection definition here? What's the preferred/best practice when mapping a collection?

File: src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "user_team")
+@org.hibernate.annotations.Immutable
+public class UserTeam {
+
+    @Embeddable
+    public static class Id implements Serializable {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java]: This should be a separate class. Also, the class name is too generic, I would advise naming it UserTeamId.

File: src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        public Long getUserId() {
+            return userId;
+        }
+
+        public Long getTeamId() {
+            return teamId;
+        }
+
+        public boolean equals(Object o) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java]: When overriding methods, please always use @Override annotation. Also, please have a look at the https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html class, it might help you write the equals and hashCode methods.

File: src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            if (o != null && o instanceof Id) {
+                Id that = (Id) o;
+                return this.userId.equals(that.userId) && this.teamId.equals(that.teamId);
+            }
+            return false;
+        }
+
+        public int hashCode() {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java]: When overriding methods, please always use @Override annotation. Also, please have a look at the https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html class, it might help you write the equals and hashCode methods.

File: src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+import javax.persistence.Embeddable;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "user_team")
+@org.hibernate.annotations.Immutable
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java]: Please do the proper import for the annotation, it's less readable with the full package qualifier.

File: src/main/java/eu/execom/labs/teamodoro/service/UserService.java [hunk 10]
```diff
@@ -21,6 +24,9 @@
     @Autowired
     TeamRepository teamRepository;
 
+    @Autowired
+    UserTeamRepository userTeamRepository;
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Why these fields are not private?

File: src/main/java/eu/execom/labs/teamodoro/service/UserService.java [hunk 11]
```diff
@@ -29,6 +35,7 @@ public Long findUserId(String googleId) {
         return null;
     }
 
+    @Transactional(readOnly = true)
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Why does it need to be transactional?

File: src/main/java/eu/execom/labs/teamodoro/service/UserService.java [hunk 12]
```diff
@@ -38,18 +45,22 @@ public void save(User user) {
     }
 
     public List<Team> findAllTeamsWithUsers(String googleId) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Why it is not transactional?

File: src/main/java/eu/execom/labs/teamodoro/service/UserService.java [hunk 13]
```diff
@@ -38,18 +45,22 @@ public void save(User user) {
...
+            team.getUserTeams().size(); // initialize (load) userTeams of the team - alternative is
+            // Hibernate.initialize(team.getUserTeams());
+            teams.add(team);
         }
         return teams;
     }
 
-    public User findOneWithTeams(String googleId) {
+    @Transactional
+    public void addTeamToUser(String googleId, Team team) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Do you add a user to the team, or team to the user? Maybe more clear method name would be *addUserToTeam*.

File: src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp [hunk 14]
```diff
@@ -55,71 +62,76 @@
...
 
-<!--     Look of the table should be improved for better readability -->
+    <!--     Look of the table should be improved for better readability -->
     <table class="table table-bordered table-hover table-striped">
       <thead>
         <tr>
-          <th><spring:message code="user.team"/></th>
-          <th><spring:message code="user.team.members"/></th>
+          <th>Team</th>
+          <th>Users</th>
```
Comment (stanislavgrujic) [inline src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp]: Is this change intentional? Why remove I18N?

File: src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp [hunk 15]
```diff
@@ -16,8 +17,10 @@
         <a class="navbar-brand" href="#">Teamodoro @Execom</a>
       </div>
       <ul class="nav navbar-nav navbar-right">
-        <li><a href="?lang=en"><span class="glyphicon glyphicon-flag"></span> English</a></li>
-        <li><a href="?lang=de"><span class="glyphicon glyphicon-flag"></span> Deutchland</a></li>
+        <li><a href="?lang=en"><span
+            class="glyphicon glyphicon-flag"></span> English</a></li>
+        <li><a href="?lang=de"><span
+            class="glyphicon glyphicon-flag"></span> Deutchland</a></li>
```
Comment (stanislavgrujic) [inline src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp]: Spelling: Deutchland -> Deutschland.
Deutschland is a country, Deutsch is a language.

File: src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp (hunk 3 above)
Comment (stanislavgrujic) [inline src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp]: Removed I18N.

File: src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp [hunk 16]
```diff
@@ -55,71 +62,76 @@
...
                     <tr>
-                      <td>${user.name}</td>
+                      <td>${userTeam.user.name}</td>
                       <td>Pomodoro timer</td>
                       <td>
                         <button type="button"
                           class="btn btn-danger btn-xs"
                           data-toggle="modal"
-                          data-target="#removeUserModal"><spring:message code="user.delete.member"/></button>
+                          data-target="#removeUserModal">Remove</button>
```
Comment (stanislavgrujic) [inline src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp]: Removed I18N.

File: src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp [hunk 17]
```diff
@@ -55,71 +62,76 @@
...
                       </td>
                     </tr>
                   </c:forEach>
                 </tbody>
               </table>
               <button type="button" class="btn btn-primary"
-                data-toggle="modal" data-target="#newTeamModal"><spring:message code="user.add.member"/></button>
+                data-toggle="modal" data-target="#newMemberModal">New
+                Member</button>
```
Comment (stanislavgrujic) [inline src/main/webapp/WEB-INF/views/jsp/Dashboard.jsp]: Removed I18N.

**Comment by stanislavgrujic** [comment]
One general remark when using the markup files. It makes it the most readable if the code is structured as:
```
<tag1>
  <tag2>Simple tag</tag>
  <tag2 attribude="Tag attribute" />
</tag1>
```

And always try to stick to this convention, unless a tag line is very long (e.g, longer than 120 chars), when it's ok to break the tag definition in two lines.
If you break the tags frequently, the file will be difficult to read and follow that tag elements.

File: src/main/java/eu/execom/labs/teamodoro/entity/Team.java (hunk 4 above)
Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: Considering the book (*Java Persistence with Hibernate - page p183*) generally we have following situation: 
> Bags have the most efficient performance characteristics of all the collections you can use for a bidirectional one-to-many entity association. By default, collections in Hibernate are loaded when they’re accessed for the first time in the application. Because a bag doesn’t have to maintain the index of its elements (like a list) or check for duplicate elements (like a set), you can add new elements to the bag without triggering the loading. This is an important feature if you’re going to map a possibly large collection of entity references.

In particular situation (`according to the book - p192`) `Set` will be more appropriate regarding the fact that there cannot be duplicate links between `Team` and `UserTeam`. However, with Set, and its recommended initialization `java.util.HashSet`, we have not ordered rendering of teams on the dashboard (each refresh different order).

One of solutions could be `SortedSet`, initialized with a `TreeSet`. In this situation we need to provide that `UserTeam` implements `Comparable` interface and to implement `compareTo` method. We can discuss about such solution and implement it in one of the next steps.

Because of the above reasons, I have decided to choose `Collection` interface with recommended initialization `ArrayList`. Teams are rendered by the order of their input time.

Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: Ok, nice explanation. Thanks. :)

Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: One remark - is the order of teams rendered by the input time, or the update time? In this case, it doesn't matter, because the rows won't be updated, but it may matter in some other queries, if the latter is the case.
I think it's better to make it explicit (e.g, by team_id asc), rather than implicit, so that it is clear how they are being fetched.

File: src/main/java/eu/execom/labs/teamodoro/service/UserService.java (hunk 12 above)
Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Absolutely correct :-) Mistake during the merging. `@Transactional` from the above method should go there.

File: src/main/java/eu/execom/labs/teamodoro/entity/Team.java (hunk 4 above)
Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: Ok. I agree, and have additional observation/question. We are rendering the teams of particular user, and, for each team, we have a list of users that should be also sorted by same logic. So, maybe solution with sorted set is not appropriate. Appropriate could be to provide in program sorting teams of particular user by `teamId`, and on other side sorting users of particular team by `userId`. Could it be solution?

File: src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java (hunk 9 above)
Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java]: The reason why there is used annotation with full path is to give an accent that we use native Hibernate annotation (not provided by JPA specification). To provide to the other code reader easy distinction between pure hibernate and jpa feature. Such convention is also recommended by the book. However, maybe we shouldn't follow this convention because Hibernate is only one part of our application and it is not necessary to put accent on it. What do you think?

Regarding Immutable in this situation - it is not obligatory to have it here. But Hibernate can do some optimizations, such avoiding dirty checking during flushing of the persistence context.

File: src/main/java/eu/execom/labs/teamodoro/entity/Team.java (hunk 4 above)
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/Team.java]: I think that if you use http://docs.oracle.com/javaee/5/api/javax/persistence/OrderBy.html you should be able to specify the list of properties by which you want to do the sorting, so you could use *@OrderBy("teamId, userId")*

File: src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java (hunk 9 above)
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/UserTeam.java]: A general remark - premature optimization is often considered more harmful than beneficial, so these kind of things should be used only if you identify it as slow and find that this annotation is helping. I'd remove it, for the moment.

File: src/main/java/eu/execom/labs/teamodoro/configuration/SpringWebInit.java (hunk 1 above)
Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/configuration/SpringWebInit.java]: I made this after poor exploring on the internet. 

Thanks to this comment, I have examined in more details *Spring Framework Reference - section 22.2 The DispatcherServlet* ([http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#mvc-servlet](url)) and *Spring in Action 4th (section 5.1.2)* and concluded that we have made mistake at the beginning. It is not good practice to put all configuration inside one config class. We have single root context (see spring reference) and in accordance with this we should have `WebConfig.java` class for presentation layer, `RootConfig.java` class for business layer and `DataConfig.java` class for repository layer. `Root.Config.java` (that will import `DataConfig.java`) will contain all the infrastructure beans. 

In servlet configuration we should put WebConfig.java and in root configuration Root.Config.java.

*The configuration classes returned from getServletConfigClasses() will define beans for DispatcherServlet’s application context. Meanwhile, the configuration classes returned getRootConfigClasses() will be used to configure the application context created by ContextLoaderListener.*

I suggest to make separation of application layers before next commit. Do you agree?

Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/configuration/SpringWebInit.java]: Yes, thank you.

File: src/main/java/eu/execom/labs/teamodoro/service/UserService.java (hunk 13 above)
Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Actually, we adding team by the user (and automatically join this user to created team - he is the first user). So appropriate name should be only addTeam. Do you agree?

Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/service/UserService.java]: Yes.

File: src/main/java/eu/execom/labs/teamodoro/entity/User.java [hunk 18]
```diff
@@ -59,20 +62,17 @@ public void setEmail(String email) {
...
-    public void setTeams(List<Team> teams) {
-        this.teams = teams;
+    public Collection<Pomodoro> getPomodoros() {
+        return pomodoros;
     }
 
-    public List<Pomodoro> getPomodoros() {
-        return pomodoros;
+    public void addPomodoro(Pomodoro pomodoro) {
+        getPomodoros().add(pomodoro);
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/User.java]: No need to call getter on own field in order to access it.

Comment (PavleP) [inline src/main/java/eu/execom/labs/teamodoro/entity/User.java]: Collection of pomodoros in user Entity has been lazy initialized. Due to this `getPomodoros()` instead of `pomodoros` , collection of pomodoros will be initialized when we will call addPomodoro method in persistent context (transaction).

*Java Persistent with Hibernate p249:*

> This is one point where Hibernate isn’t completely  transparent, but it’s good practice anyway to use getter methods instead of direct instance variable access.
